activate.c

```diff
@@ -72,17 +72,24 @@ static void activate_mapping(struct irq_info *info, void *data __attribute__((un
 	sprintf(buf, "/proc/irq/%i/smp_affinity", info->irq);
 	file = fopen(buf, "w");
 	if (!file)
-		return;
+		goto error;
 
 	cpumask_scnprintf(buf, PATH_MAX, applied_mask);
 	ret = fprintf(file, "%s", buf);
-	if (ret < 0 || fflush(file)) {
-		log(TO_ALL, LOG_WARNING, "cannot change irq %i's affinity, add it to banned list", info->irq);
-		add_banned_irq(info->irq);
-		remove_one_irq_from_db(info->irq);
+	errsave = errno;
+	if (fclose(file))
+		goto error;
+	if (ret < 0) {
+		errno = errsave;
+		goto error;
 	}
-	fclose(file);
 	info->moved = 0; /*migration is done*/
```

Comment:
With the code change, only non-error irq's info->moved is assigned 0. Previously all info->moved which reached here will be 0. I'm unsure of the correctness. The rest of the patch is good to me.

Reply:
That's a good point. I would assume that **not** marking the IRQ as "moved" uppon error is correct. However, there may be side effects with doing this.

Reply:
There is no record of the irq being moved, as the actual state of the irqs will be re-determined next time /proc/interrupts is parsed.  The only question is, should we ban an irq just because we werent able to set its affinity.  APIC's being out of space is a transient issue, banning an irq is permanent for the life of the process.  This patch looks fine, but I'd just log the error as an administrative warning, and let irqbalance try again later by _not_ calling add_banned_irq and remove_one_irq_from_db

Reply:
maybe I should do a separate commit that removes `add_banned_irq()` and `remove_one_irq_from_db()` then?

Reply:
I'd be ok with that

Reply:
done :+1:

Reply:
But isn't that a direct revert of this commit? 55c5c321c73e4c9b54e041ba8c7d542598685bae

Reply:
Yes, save for the log message indicating that affinity setting went wrong, which I think should be enough to inform an admin that there was a transient error, or that they need to take action to be more direct in their affinity setting policy.  Or do you disagree?

Reply:
On aarch64, there are interrupts that are local to every CPU and that **cannot** be pinned elsewhere. Here is such an example on my raspberry pi 4B:

```
LibreELEC:~ # cat /proc/interrupts 
           CPU0       CPU1       CPU2       CPU3       
 11:    2917915    6446201    6099729    4041827     GICv2  30 Level     arch_timer
 14:    5081500          0          0          0     GICv2  65 Level     fe00b880.mailbox
 15:        342          0          0          0     GICv2 112 Level     DMA IRQ
 16:      30922          0          0          0     GICv2 114 Level     DMA IRQ
 17:          0          0          0          0     GICv2 116 Level     DMA IRQ
 22:          0          0          0          0     GICv2 122 Level     DMA IRQ
 23:          0          0          0          0     GICv2 123 Level     DMA IRQ
 24:          0          0          0          0     GICv2  48 Level     arm-pmu
 25:          0          0          0          0     GICv2  49 Level     arm-pmu
 26:          0          0          0          0     GICv2  50 Level     arm-pmu
 27:          0          0          0          0     GICv2  51 Level     arm-pmu
 32:          0          0          0          0     GICv2 175 Level     PCIe PME, aerdrv
 33:     696011          0          0          0     GICv2 129 Level     vc4 hvs
 34:     112404          0          0          0     GICv2 189 Level     eth0
 35:      14039          0          0          0     GICv2 190 Level     eth0
 36:        154          0          0          0  BRCM STB PCIe MSI 524288 Edge      xhci_hcd
 37:      83388          0          0          0     GICv2  66 Level     VCHIQ doorbell
 38:       7482          0          0          0     GICv2 153 Level     uart-pl011
 39:          1          0          0          0  interrupt-controller@7ef00100   4 Edge      vc4 hdmi hpd connected
 40:          1          0          0          0  interrupt-controller@7ef00100   5 Edge      vc4 hdmi hpd disconnected
 41:         41          0          0          0  interrupt-controller@7ef00100   1 Edge      vc4 hdmi cec rx
 42:       2090          0          0          0  interrupt-controller@7ef00100   0 Edge      vc4 hdmi cec tx
 43:          0          0          0          0  interrupt-controller@7ef00100  10 Edge      vc4 hdmi hpd connected
 44:          0          0          0          0  interrupt-controller@7ef00100  11 Edge      vc4 hdmi hpd disconnected
 45:          0          0          0          0  interrupt-controller@7ef00100   7 Edge      vc4 hdmi cec rx
 46:          0          0          0          0  interrupt-controller@7ef00100   8 Edge      vc4 hdmi cec tx
 47:          0          0          0          0     GICv2 107 Level     fe004000.txp
 48:          0          0          0          0     GICv2 141 Level     vc4 crtc
 49:          0          0          0          0     GICv2 142 Level     vc4 crtc, vc4 crtc
 50:     696017          0          0          0     GICv2 133 Level     vc4 crtc
 51:          0          0          0          0     GICv2 138 Level     vc4 crtc
 52:         60          0          0          0     GICv2 149 Level     fe804000.i2c
 53:      44651          0          0          0     GICv2 158 Level     mmc1, mmc0
 54:        834          0          0          0     GICv2 106 Level     v3d
 55:          0          0          0          0     GICv2 130 Level     feb10000.codec
IPI0:       552       3683       3655       1531       Rescheduling interrupts
IPI1:    510105    1731674    1594065    1236233       Function call interrupts
IPI2:         0          0          0          0       CPU stop interrupts
IPI3:         0          0          0          0       CPU stop (for crash dump) interrupts
IPI4:         0          0          0          0       Timer broadcast interrupts
IPI5:     33753      78819      76008      52447       IRQ work interrupts
IPI6:         0          0          0          0       CPU wake-up interrupts
Err:          0
LibreELEC:~ # grep . /proc/irq/*/smp_affinity_list
/proc/irq/1/smp_affinity_list:0-3
/proc/irq/10/smp_affinity_list:0-3
/proc/irq/11/smp_affinity_list:0-3
/proc/irq/12/smp_affinity_list:0-3
/proc/irq/13/smp_affinity_list:0-3
/proc/irq/14/smp_affinity_list:0-3
/proc/irq/15/smp_affinity_list:0-3
/proc/irq/16/smp_affinity_list:0-3
/proc/irq/17/smp_affinity_list:0-3
/proc/irq/2/smp_affinity_list:0-3
/proc/irq/22/smp_affinity_list:0-3
/proc/irq/23/smp_affinity_list:0-3
/proc/irq/24/smp_affinity_list:0
/proc/irq/25/smp_affinity_list:1
/proc/irq/26/smp_affinity_list:2
/proc/irq/27/smp_affinity_list:3
/proc/irq/3/smp_affinity_list:0-3
/proc/irq/32/smp_affinity_list:0-3
/proc/irq/33/smp_affinity_list:0-3
/proc/irq/34/smp_affinity_list:0-3
/proc/irq/35/smp_affinity_list:0-3
/proc/irq/36/smp_affinity_list:0-3
/proc/irq/37/smp_affinity_list:0-3
/proc/irq/38/smp_affinity_list:0-3
/proc/irq/39/smp_affinity_list:0-3
/proc/irq/4/smp_affinity_list:0-3
/proc/irq/40/smp_affinity_list:0-3
/proc/irq/41/smp_affinity_list:0-3
/proc/irq/42/smp_affinity_list:0-3
/proc/irq/43/smp_affinity_list:0-3
/proc/irq/44/smp_affinity_list:0-3
/proc/irq/45/smp_affinity_list:0-3
/proc/irq/46/smp_affinity_list:0-3
/proc/irq/47/smp_affinity_list:0-3
/proc/irq/48/smp_affinity_list:0-3
/proc/irq/49/smp_affinity_list:0-3
/proc/irq/5/smp_affinity_list:0-3
/proc/irq/50/smp_affinity_list:0-3
/proc/irq/51/smp_affinity_list:0-3
/proc/irq/52/smp_affinity_list:0-3
/proc/irq/53/smp_affinity_list:0-3
/proc/irq/54/smp_affinity_list:0-3
/proc/irq/55/smp_affinity_list:0-3
/proc/irq/6/smp_affinity_list:0-3
/proc/irq/7/smp_affinity_list:0-3
/proc/irq/8/smp_affinity_list:0-3
/proc/irq/9/smp_affinity_list:0-3
LibreELEC:~ # echo 2 > /proc/irq/26/smp_affinity_list
sh: write error: Input/output error
```

Reading again through commit https://github.com/Irqbalance/irqbalance/commit/55c5c321c73e4c9b54e041ba8c7d542598685bae, it seems like the goal was to prevent such errors from happening repeatedly. But it had unfortunate side-effects on x86_64.